app/visualisation_methods/dendrogram/plugin.rb

```diff
@@ -85,42 +89,57 @@ def get_the_best_depth(data, root, hash)
+      {"id" => root, "depth" => depth, "children" => [{"id" => -d[0]}, {"id" => -d[1]}]}
     elsif d[0] < 0 && d[1] > 0
-      "{\"id\":\"#{node}\",\"children\":[#{create_json(hash, d[1])},{\"id\":\"#{-d[0]}\"}]}"
+      {"id" => root, "depth" => depth, "children" => [create_hash(data, d[1], depth+1), {"id" => -d[0]}]}
     elsif d[0] > 0 && d[1] < 0
-      "{\"id\":\"#{node}\",\"children\":[#{create_json(hash, d[0])},{\"id\":\"#{-d[1]}\"}]}"
+      {"id" => root, "depth" => depth, "children" => [create_hash(data, d[0], depth+1), {"id" => -d[1]}]}
     elsif d[0] > 0 && d[1] > 0
-      "{\"id\":\"#{node}\",\"children\":[#{create_json(hash, d[0])},#{create_json(hash, d[1])}]}"
+      {"id" => root, "depth" => depth, "children" => [create_hash(data, d[0], depth+1), create_hash(data, d[1], depth+1)]}
     end if d!=nil
   end
 
   ##
-  # create hash with strusture of dendrogram with depth {id => 1, depth => 1, children => [{id, depth, children => []}]}
-  def create_hash(hash, node, depth)
-    d = hash[node.to_s]
+  # get depth of dendrogram with max count of leafs equal 250, data = {'1'=>[-2,3],'3'=>[-4,-5]}
+  # hash = {"id"=>1, "depth"=>0, "children"=>[{"id"=>3, "depth"=>1, "children"=>[{"id"=>4}, {"id"=>5}]}, {"id"=>2}]}
+  # return integer
+  def get_the_best_depth(data, root, hash)
+    max_depth = get_max_depth(data, root, 0)
+    (0..max_depth).each do |depth|
+      if count_of_leafs(hash, depth, 0).to_i > 250
+        return depth-1
+      end
+    end
+    return max_depth
+  end
+
+  ##
+  # return maximum depth of dendrogram, data = {'1'=>[-2,3],'3'=>[-4,-5]}
+  def get_max_depth(data, root, depth)
+    d = data[root.to_s]
     if d[0] < 0 && d[1] < 0
-      {"id" => node, "depth" => depth, "children" => [{"id" => -d[0]}, {"id" => -d[1]}]}
+      depth+1
     elsif d[0] < 0 && d[1] > 0
-      {"id" => node, "depth" => depth, "children" => [create_hash(hash, d[1], depth+1), {"id" => -d[0]}]}
+      [get_max_depth(data, d[1], depth+1), depth].max
     elsif d[0] > 0 && d[1] < 0
-      {"id" => node, "depth" => depth, "children" => [create_hash(hash, d[0], depth+1), {"id" => -d[1]}]}
+      [get_max_depth(data, d[0], depth+1), depth]
     elsif d[0] > 0 && d[1] > 0
-      {"id" => node, "depth" => depth, "children" => [create_hash(hash, d[0], depth+1), create_hash(hash, d[1], depth+1)]}
+      [get_max_depth(data, d[0], depth+1), get_max_depth(data, d[1], depth+1)].max
     end if d!=nil
   end
 
   ##
-  # get count of leafs if depth of dendrogram = depth
-  def count_of_leafs(tree, depth, count)
-    if tree['depth'] != nil
-      if tree['depth'] <= depth
-        if tree['children'].kind_of?(Array)
-          count = count + count_of_leafs(tree['children'][0], depth, count) + count_of_leafs(tree['children'][1], depth, count)
+  # get count of leafs (clusters and simulations) if depth of dendrogram = depth, hash = {'id'=>1, 'depth'=>0, 'children'=>[{'id'=>2}, {'id'=>3}]}
+  def count_of_leafs(hash, depth, count)
```

Comment:
@khalska, @Andriu01 - czy tutaj nie ma czasami redundancji :)? (wszystkie else robią to samo, a można było zrobić ifa z wieloma warunkami).  Jeśli to pisał @Andriu01 to w ramach tego pull requesta nie trzeba poprawiać...